src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -237,7 +236,21 @@ public boolean isForceExecution() {
         @Override
         public void messageReceived(final ReplicaOperationRequest request, final TransportChannel channel) throws Exception {
             try {
-                shardOperationOnReplica(request);
+                IndexShard indexShard = indicesService.indexServiceSafe(request.shardId.getIndex()).shardSafe(request.shardId.id());
+                boolean counterInc = false;
+                try {
+                    counterInc = indexShard.beforeWriteOperation();
```

Comment:
I think we can just move this all out of the try block and then we don't need to assign this to any value since int he finally block we simply decrement the counter. this would simplify this a lot I think